### ISSUE TITLE
MsalException when calling GetAccountsAsync with newer version of Microsoft.Identity.Client

### ISSUE THREAD

**Issue by twentytwokhz** [Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2#28]
### This issue is for a: (mark with an `x`)
```
- [X] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Minimal steps to reproduce
>application.GetAccountAsync(accountIdentifier) returns null and as such a MsalException is thrown saying that the account cannot be null when calling **AcquireTokenSilentAsync**

### Any log messages given by the failure
>Error thrown at [line 281 ](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/938c0ddec89e6b9105ad2ef133e2bb33a75a1e71/Extensions/TokenAcquisition.cs#L279)

### Expected/desired behavior
>should work as with nuget package Microsoft.Identity.Client 2.3.1-preview (in this one there is no exception)

### OS and Version?
> Windows 10
### Versions
>Branch aspnetcore2-2-signInAndCallGraph with nuget package Microsoft.Identity.Client 2.5.0-preview

**Comment by jmprieur** [comment]
Thanks @twentytwokhz .
So this would be regression with MSAL.NET?
I'll investigage. cc: @henrik-me @jennyf19 @trwalke @bgavrilMS

**Comment by twentytwokhz** [comment]
I think it is..since the problem appears after the call to GetAccountsAsync using the correct accountid. Worked in 2.3.1, not working with latest. Should have reported it over there, sorry :)

**Comment by jmprieur** [comment]
@twentytwokhz do you have repro steps that I could use? this does not seem systematic. would it be with guest accounts (accounts invited into the tenant) ?

**Comment by twentytwokhz** [comment]
I'm working with accounts created inside the tenant, so organizational accounts only. Some repo steps below:
1. get repo with branch **aspnetcore2-2-signInAndCallGraph** from https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2
2. enter the required settings in appsettings.json for a single tenant app with only organizational accounts
3. set nuget package Microsoft.Identity.Client to 2.5.0-preview
4. login with an AD account from that tenant => will work fine
5. enter on the contact page which uses the graph api
6. MsalException will be thrown

**Comment by jmprieur** [comment]
thanks @twentytwokhz 
Does it fail with any account from that tenant? this seems to work fine for me, except for guest accounts. But I'll keep looking at it.

**Comment by twentytwokhz** [comment]
@jmprieur yes, it fails with any AD account in the tenant: test@tenant.onmicrosoft.com, etc
then the exception is caught inside the catch MsalUiRequiredException in the controller and then it's an infinite loop of exceptions

**Comment by jmprieur** [comment]
@twentytwokhz : any chance this could be related to #30 ?
- did you allow cookies in your browser? 
- does it also repro with our release candidate: https://www.myget.org/feed/aad-clients-nightly/package/nuget/Microsoft.Identity.Client ?

**Comment by twentytwokhz** [comment]
@jmprieur never tried with a guest account
1. yes
2. tried it now with 2.6.0 and still reproduces

what i currently have modified from latest code on the branch  aspnetcore2-2-signInAndCallGraph:
`if (jwtToken == null)` replaced by `if (tenantId == null)` on line 34 AadIssuerValidator.cs and
`app = new ConfidentialClientApplication(_azureAdOptions.ClientId, currentUri, credential, userTokenCache, null);` becomes
```
 string authority = _azureAdOptions.Instance + _azureAdOptions.TenantId;
            app = new ConfidentialClientApplication(_azureAdOptions.ClientId, authority, currentUri, credential, userTokenCache, null);
``` on line 242 TokenAcquisition.cs

basically just some minor local fixes.

**Comment by jmprieur** [comment]
Thanks @twentytwokhz 
I've investigated what happened, and it turns out that something in MSAL (getting the OIDC metadata) made ASP.NET Core change the session, and therefore the session based token cache was empty (as the session was different)

I've replaced the SessionBasedTokenCache, by an InMemoryToken cache, and this is working very well.

Please confirm this works for you

**Comment by twentytwokhz** [comment]
This seems to work fine now. Thank you @jmprieur !

**Comment by jmprieur** [comment]
Thanks for confirming, @twentytwokhz 
I've also found it way quicker :)

**Comment by Prabu76** [comment]
Hi,
I am using latest version of MSAL 4.14.0. When I try to call below code:
IPublicClientApplication clientApp = PublicClientApplicationBuilder
                                                                    .Create(ApplicationId)
                                                                    .WithAuthority(AuthorityUrl)
                                                                    .Build();
                var userAccounts = await clientApp.GetAccountsAsync();

It is always returning zero. Not sure what is going on.

**Comment by TiagoBrenck** [comment]
Hi @Prabu76 

I have a question. Why would you use IPublicClientApplication on a web scenario? It should be IConfidentialClientApplication.

But regardless that, if you are looking to return the cached tokens from MSAL, you can:
   - Or, have a singleton of you ClientApplication, which you would leverage the default object memory cache from MSAL (if you build a new ClientApp or the singleton obj get cleared, you lose all the caches).
   - Or, keep building a new ClientApplication when needed, but you must then implement a token cache (we provide a bunch of options on Microsoft.Identity.Web package) and hook this token cache to the new obj built.

Thanks

**Comment by Prabu76** [comment]
Hi,
I tried to use the existing developer samples provided here https://github.com/Microsoft/PowerBI-Developer-Samples
I have downloaded the code and tried with ID's... This is the code that power BI documentation point to as reference. I would appreciate if any other sample code to access and get token from Azure AD app.

**Comment by Prabu76** [comment]
@TiagoBrenck  ConfidentialClientApplication requires clientsecret. I have registered my application as Native app which have only clientID.
Do you have any sample code to show the options you mentioned

**Comment by TiagoBrenck** [comment]
Oh ok, for native app is public indeed. 
You can take a look at this native app implementation: https://github.com/Azure-Samples/active-directory-dotnet-native-aspnetcore-v2/tree/master/1.%20Desktop%20app%20calls%20Web%20API/TodoListClient

We have a class there called TokenCacheHelper, that you might be interested in

**Comment by jmprieur** [comment]
@Prabu76 : you can even use a better token cache serialization (cross platform, productized, battle tested), See https://www.nuget.org/packages/Microsoft.Identity.Client.Extensions.Msal and the doc:
https://github.com/AzureAD/microsoft-authentication-extensions-for-dotnet/wiki/Cross-platform-Token-Cache

cc: @TiagoBrenck

**Comment by Prabu76** [comment]
@TiagoBrenck  @jmprieur Thanks for the response.
Before doing this for the below code:

IPublicClientApplication clientApp = PublicClientApplicationBuilder
.Create(ApplicationId)
.WithAuthority(AuthorityUrl)
.Build();
var userAccounts = await clientApp.GetAccountsAsync();

I am also checking to see if I need to grant any permission to the application registered in Azure. I just registered the application in Azure AD. And gave access rights to power BI reports and datasets. I am not sure if any other Grant Permissions needed. I read some documentation saying Implicit Grant. Not sure.

**Comment by Prabu76** [comment]
@TiagoBrenck @jmprieur When I tried to call using below code:

AuthenticationContext authContext = new AuthenticationContext(AuthorityUrl);
var user = new UserCredential(Username,Password);
string token = authContext.AcquireToken(ResourceUrl, ApplicationId, user).AccessToken;

I am getting an error - 
unknown_user_type: Unknown User Type
Exception Details: Microsoft.IdentityModel.Clients.ActiveDirectory.AdalException: unknown_user_type: Unknown User Type

Is this that I am missing something from permission point of view?

**Comment by TiagoBrenck** [comment]
@Prabu76 your code is using ADAL library (Microsoft.IdentityModel.Clients.ActiveDirectory).

Could you update it to MSAL library (Microsoft.Identity.Client) so you could benefit from our sample code?
